### PR TITLE
adagrams

### PR DESCRIPTION
**Pull request by unkcodesquick**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | method signature, parameters, method block 
What are the advantages of using git when collaboratively working on one code base? | ability to track changes and work off different machines.
What kind of relationship did you and your pair have with the unit tests? | lower engagement than expected, but they were very helpful guiding us - good to see how they were structured to help guide thought process for creating methods.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? |select, reduce, min_by- they were helpful to parse through sets of information and return a single value. they helped dry up our code.
What was one method you and your pair used to debug code? | we used pry - to look at what variables/data types to see if they matched our plans/expectations
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | How our code hygiene has decreased since we've learned more code - we wished we did more whiteboarding and psuedocode instead of going straight to writing code.

### CODE REVIEW

**Comment by dHelmgren** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Good answers
Both teammates contributed to the codebase | YEAH!
Small commits with meaningful commit messages | Good, small, well spoken commits!
**Code Requirements** |
`draw_letters` method | Besides the issue below, well structured and readable.
Uses appropriate data structure to store the letter distribution | Your hash actually doesn't do the appropriate task!
All tests for `draw_letters` pass | Pass!
`uses_available_letters?` method | It works, though you've created a little more work for yourself and for the computer than you needed to. I don't think that converting the arrays to hashes actually helps performance or clarity.
All tests for `uses_available_letters?` pass | P-P-P-Pass!
`score_word` method | See comments. This works, but gosh it's not very clear what's happening. 
Uses appropriate data structure to store the letter scores | I do think that using a fixed value hash for the scoring scheme is a great idea!
All tests for `score_word` pass | Pass!
`highest_score_from` method | Oooh, this is full of spiffy one liners! Do be aware of using enumerables too much though. Each enumerable hides a loop from you.
Appropriately handles edge cases for tie-breaking logic | This was a very clever and readable solution.
All tests for `highest_score_from` pass | Pass!
**Overall** | Your code all works correctly, however, it seems like you tried to use the same tool to solve most problems. Hashes are really great for some of the things you are trying to do, but arrays are a better option in as many cases. Let me know if you want some help understanding why some of these situations called for different data structures.

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
+require "pry"
+SCORE_TO_LETTERS = {1 => ["a", "e", "i", "o", "u", "l", "n", "r", "s", "t"], 2 => ["d", "g"], 3 => ["b", "c", "m", "p"],
+  4 => ["f", "h", "v", "w", "y"], 5 => ["k"], 8 => ["j", "x"], 10 => ["q", "z"]}
+LETTERS_TO_QUANTITY = {"a" => 9, "b" => 2, "c" => 2, "d" => 4, "e" => 12, "f" => 2, "g" => 3, "h"=> 2, "i" => 9, "j" => 1, "k" => 1,
+    "l" => 4, "m" => 2, "n" => 6, "o" => 8, "p" => 2, "q"=> 1, "r" => 6, "s" => 4, "t" => 6, "u"=> 4, "v" => 2,
+    "w" => 2, "x" => 1, "y"=> 2, "z" => 1}
+def draw_letters
+  letter_bank = Array.new
+  10.times do
+    random_letter_picked = LETTERS_TO_QUANTITY.keys.sample
```
Comment (dHelmgren) [inline lib/adagrams.rb:10]: So, you have a probability problem here. When you sample from the keys, the odds are not weighted by the number of each tile in the `letter_bank`.  Your results give a 1/26 chance to draw any letter, but that's not accurate is it?

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  letter_to_quantity = Hash.new(0)
+  letters.each do |letter|
+    letter_to_quantity[letter] +=1
+    end
+ return letter_to_quantity
+end
+
+def uses_available_letters?(user_input_word, letter_bank)
+  array_of_input_letters = user_input_word.split('')
+  letter_bank_hash = letters_to_hash(letter_bank)
```
Comment (dHelmgren) [inline lib/adagrams.rb:32]: Is there any reason you couldn't have done this with 2 arrays?

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  user_input_letters_hash = letters_to_hash(array_of_input_letters)
+  output = true
+  user_input_letters_hash.each do |key,value|
+    if !letter_bank_hash.keys.include?(key) # look at looping of outputs, see what happens if use
```
Comment (dHelmgren) [inline lib/adagrams.rb:36]: As soon as one of these is false, do you need to check the whole hash still?

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  if user_input_word.length == 0
+    total_score = 0
+  else
+    array_of_input_letters = user_input_word.downcase.split("")
+    user_input_word_hash = letters_to_hash(array_of_input_letters)
+      array_of_letter_scores = []
+      user_input_word_hash.each do |word_key, word_value|
+        SCORE_TO_LETTERS.each do |score_key, score_value|
+          if score_value.include?(word_key)
+            score = score_key*word_value
```
Comment (dHelmgren) [inline lib/adagrams.rb:55]: It took me a minute to see why this works. Again, using a hash here is actually more work for you, and hindering clarity.

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            array_of_letter_scores.push(score)
+          end
+        end
+      end
+      total_score =  array_of_letter_scores.reduce(:+)
```
Comment (dHelmgren) [inline lib/adagrams.rb:60]: `reduce` is hiding a loop from you, though you could have used reduce to solve this problem by replacing an earlier `.each`.

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        total_score += 8
+      end
+      return total_score
+    end
+  end
+#played_words = ['BBBBBB', 'AAAAAAAAAA']
+
+def highest_score_from(played_words)
+  collection_of_played_words_and_scores = []#array of hashes
+  played_words.each do |word|
```
Comment (dHelmgren) [inline lib/adagrams.rb:71]: nice! I love this, very clear!

File: lib/adagrams.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    played_word_to_score = {} #how to deal with word duplicates
+    played_word_to_score[:word] = word
+    played_word_to_score[:score] = score_word(word)
+    collection_of_played_words_and_scores.push(played_word_to_score)
+  end
+  top_score = collection_of_played_words_and_scores.reduce(0){ |memo, h| h[:score] > memo ? h[:score] : memo } # ==> {score: n}
```
Comment (dHelmgren) [inline lib/adagrams.rb:77]: Yum, this is also great!
